docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,130 @@
+# Syncing LDAP Groups to OpenShift Groups
+
+OpenShift cluster admins need a way to sync their OpenShift `Groups` to external records in an LDAP server. This requires that we have a way of retrieving the LDAP users that are the members of a group, determining the OpenShift `Group` name for this collection, and updating the OpenShift record, removing the record as necessary when the corresponding LDAP record is removed. Necessary information about the LDAP server set-up and schema will be provided by the user in a `yaml` configuration file and referenced in the `oadm sync-groups` command.
+
+## Use Cases
+1. I want to sync all LDAP groups minus some blacklist to OpenShift `Groups`
+* I want to sync a specific set of LDAP groups to OpenShift `Groups`
+* I want to sync all OpenShift groups minus some blacklist with the LDAP server
+* I want to sync specific OpenShift `Groups` with the LDAP server records
+* I want to remove OpenShift `Groups` that were previously backed by LDAP records that no longer exist
+* I want to specify an explicit mapping from LDAP group identifiers to OpenShift `Group` resources
+
+## Supported LDAP Group Schemas
+1. groups as first-class LDAP entries that contain lists of members of the group
+2. users as first-class LDAP entries that contain lists of groups they are a member of
+3. schema 2 above, also including first-class groups to store metadata
+
+## Required OpenShift `Group` Metadata
+OpenShift `Group`'s `Annotations` will be used to store metadata regarding the sync process. `Annotations` will store:
+
+1. LDAP unique identifier (`openshift.io/ldap-uid`)
+* last sync time (`openshift.io/ldap-last-sync-time`)
+* if available, the LDAP resource version (`openshift.io/ldap-resource-version`).
+
+OpenShift `Group`'s `Labels` will also be used to store metadata regarding the sync process. `Labels` will store:
+
+1. LDAP server URL (`openshift.io/external-record-provider-url`).
+
+## Data Flow
+1. Determine ordered list of LDAP groups to be sync
+* Collect LDAP members and metadata
+* Determine OpenShift `Identities` from LDAP group member entries
+* Determine OpenShift `Users` from OpenShift `Identities`
+* Determine OpenShift `Group` from LDAP group metadata
+* Populate new OpenShift `Group`'s `Users` from resulting OpenShift `UserIdentityMappings`, update OpenShift `Group` metadata fields tied to LDAP attributes, leave other OpenShift `Group` fields unchanged
+
+##  Determining Ordered List of LDAP Groups To Sync
+An `LDAPGroupLister` determines what LDAP groups needs to be synced and outputs the result as a set of unique identifier strings (called "LDAP group UIDs" in this document). The other objects that take LDAP group UIDs need to understand the format of this string (e.g. these objects will be tightly coupled). For example, an `LDAPGroupLister` for schema 1 above cannot be used with a `LDAPGroupDataExtractor` for schema 2. The four `LDAPGroupLister` implementations that are necessary are:
```

Comment:
actually, you can't assume a DN is stable... it is unique, but a DN can change over the lifetime of an entry, so you need to be told what to use for the LDAP group uid in the case of a group entry